### ISSUE TITLE
support namespacing

### ISSUE THREAD

**Issue by spaceone** [circuits/circuits#20]
circuits could support namespace packages to allow third party components to be easier integrated.

See http://peak.telecommunity.com/DevCenter/setuptools#namespace-packages
See also commit c6b73daf89ec16939425af2a48b6f3b2b897ba2b where we tried it in the past.

With a little adjustment it works for me with Python2.7 and every combination of the order in sys.path.

**Comment by prologic** [comment]
I'm keen to give this a try again and will do some local testing of my own :) Well done :) Let's see if we can move this forward and merge it in for 3.2?

**Comment by prologic** [comment]
BTW: There's a bug with the implementation of this. If you run anything with circuits under PyPy there's a recursion problem with the namespace code and it blows the stack :/

**Comment by prologic** [comment]
We should add documentation around this:

* [ ] How to add to the ``circuits`` namespace
* [ ] How to publish new PyPi ``circuits.*`` packages

**Comment by prologic** [comment]
bump

**Comment by spaceone** [comment]
https://www.python.org/dev/peps/pep-0420/

**Comment by prologic** [comment]
Can you TL;DR this for us?

**Comment by prologic** [comment]
Been able to reproduce that recursion bug:

```#!bash
(circuits)
prologic@daisy
Thu Jun 18 00:43:35
~/circuits
$ PYTHONPATH=. pypy ./bin/circuits.bench -e 1000000^C
...
  File "/home/prologic/work/circuits/circuits/six.py", line 331, in exec_
    exec("""exec code in globs, locs""")
  File "<string>", line 1, in <module>
  File "/home/prologic/work/circuits/circuits/__init__.py", line 38, in <module>
    exec_(fd, globals())
  File "/home/prologic/work/circuits/circuits/six.py", line 331, in exec_
    exec("""exec code in globs, locs""")
  File "<string>", line 1, in <module>
RuntimeError: maximum recursion depth exceeded
```

**Comment by spaceone** [comment]
Okay, i created PR #101 for this.
It's an error if you have two different circuits in your PYTHONPATH. (or a wrong configured namespacing package).

**Comment by prologic** [comment]
:+1:

**Comment by spaceone** [comment]
Please have a look at: https://github.com/spaceone/circuits/blob/9f1949ec13f0ff2ac2c2b0449c8be95e25aeb9c0/docs/source/dev/namespacing.rst

**Comment by prologic** [comment]
Looks good so far :+1: -- Incomplete?

**Comment by prologic** [comment]
Where are we up to with this one?

**Comment by spaceone** [comment]
Docs for your wished "How to publish new PyPi circuits.* packages" is missing. I don't know what to write there. Otherwise this seems fine.
If we want to support namespacing in circuits.protocols, etc. we have to copy that code into these __init__.py files, too.
Python3.X supports namespacing OOTB for packages which DON'T have a __init__.py. We could add a circuits.thirdparty. namespace for python3 only (bad idea imho).

**Comment by prologic** [comment]
Yeah let's try to get this over the line I think :)
(I'll try to do some more testing)

I'll see if I can create an example namespace?

Do you think it's a good idea to namespace
app, net, protocols, tools and web?

cheers
James


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by prologic** [comment]
Interesting to read the [``flask.ext`` docstring](https://github.com/mitsuhiko/flask/blob/0.10.1/flask/ext/__init__.py) -- Makes me thingk we *should not* go down this road from lessons learned before :)

**Comment by prologic** [comment]
See: mitsuhiko/flask#1484

**Comment by apollo13** [comment]
pep-0420 is the only namespacing to support. setuptools namespaces can break horribly in certain situations (ie removal of packages etc…).

I think this should be closed.

**Comment by spaceone** [comment]
hm. my projects like circuits.http are based on the current way.
if we want to implement pep0420 we need to remove every `__init__` file.

**Comment by apollo13** [comment]
@spaceone That would only be an option on py3. That said, I do not think every package needs to be namespacable. I guess if you need it, we can leave the existing support -- though I would recommend moving to circuits_http or so given the problems namespacing can cause.

Btw: what does `circuits.http` do and why is it not part of circuits ^^

**Comment by spaceone** [comment]
We should leave it as is, it works fine for me and i don't like _ in my own package name, i rather would fork circuits if this gets removed or rename the project into something else.
`circuits.http` is a web framework which all my websites run on, it's architecture forces you to be HTTP compliant, its API is still in development and it gives you full control over everything, so there are no artificial limitations (e.g. if you want to use unix sockets, if you want to give own error messages or HTML sites if problems with SSL occur, etc.). My goal is that you can really easily build truly RESTful applications with it without the need for code redundancy and workaround code (because HTTP is the uniform interface not the programming language!).
`circuits.http` is not part of circuits because I didn't speak about this with prologic and it depends on my external library https://github.com/spaceone/httoop which is python2.7 only (because python3 sucks with bytestrings). Prologic wants to enhance circuits.web (or be backwards compatible), I don't want this as I dislike circuits.web much and don't have the time to fix everything there.

**Comment by apollo13** [comment]
Interesting, re "python3 sucks with bytestrings" -- is that pre or post PEP461?

**Comment by spaceone** [comment]
well, pre-PEP4.6.1 / python 3.5. But still there are probably some things which I dislike. See https://github.com/spaceone/httoop/issues/1
I will see, maybe I can have a look in the next weeks.

**Comment by apollo13** [comment]
Btw regarding the current usage of namespacing in circuits (copied from http://peak.telecommunity.com/DevCenter/setuptools#namespace-packages):
```
You must NOT include any other code and data in a namespace package's __init__.py. Even though it may appear to work during development, or when projects are installed as .egg files, it will not work when the projects are installed using "system" packaging tools -- in such cases the __init__.py files will not be installed, let alone executed.
```

`circuits` currently has a few imports in there, so as soon as a namespaced package is installed, those imports can break :/

**Comment by prologic** [comment]
This is correct and one of the issues I have with supporting namespaces.

All of a sudden once you install another package in the same namespace; all
the "convenient" imports are gone and you can no longer just do:

from circuits import Component, Event


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by spaceone** [comment]
hmm, maybe we could hack by assignign sys.modules['circuits'].__dict__['Component'] = 'Component' :D:D:D:D

**Comment by prologic** [comment]
OH my god! That's brilliant! Also awful :)

Is there no way to support namespaces without breaking these imports and
not hacking sys.modules? :)


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by apollo13** [comment]
I am sure you are joking, but I do not think
```
sys.modules['circuits'].__dict__['Component'] = Component
```
would actually work. You'd need to have some location to actually put that code… `circuits/__init__.py` is not the one you can use since that is not even guaranteed to get loaded…

**Comment by spaceone** [comment]
if circuits has a `__init__.py` file it's not possible to make a namespace package under circuits.contrib, right!?

**Comment by spaceone** [comment]
yes, it was more a joke, but this was a stupid decision by the python guys...

**Comment by prologic** [comment]
Why can't we make circuits.contrib the namespace? I'm okay with this
approach if we do that?


James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au
